lint/queue.py

```diff
@@ -1,116 +1,70 @@
-
-                for view_id, (timestamp, delay) in last_runs.copy().items():
-                    # Lint the view if we have gone past the time
-                    # at which the lint wants to run.
-                    if time.monotonic() > timestamp + delay:
-                        self.last_runs[view_id] = time.monotonic()
-                        del last_runs[view_id]
-                        self.lint(view_id, timestamp)
-
-            except Exception:
-                util.printf('error in SublimeLinter daemon:')
-                util.printf('-' * 20)
-                util.printf(traceback.format_exc())
-                util.printf('-' * 20)
+        # If we have a newer hit than this one, we debounce
+        if last_hit > hit_time:
+            print('WORKER DELAY', vid, hit_time)
+            queue_lint(vid, last_hit - hit_time, callback)
+            return
 
-    def hit(self, view):
-        """Add a lint request to the queue, return the time at which the request was enqueued."""
-        timestamp = time.monotonic()
-        self.q.put((view.id(), timestamp, self.get_delay(view)))
-        return timestamp
+        print('WORKER EXEC', vid, hit_time)
+        callback(vid, hit_time)
 
-    def delay(self, milliseconds=100):
-        """Add a millisecond delay to the queue."""
-        self.q.put(milliseconds / 1000.0)
+    # We cannot fire rapidly `set_timeout_async` or Sublime will miss some
+    # events eventually. So we use a membership test `vid in last_seen` to tell
+    # if we already queued an async worker or not.
+    if vid not in last_seen:
+        print('QUEUE LINT', vid, hit_time, delay * 1000)
+        sublime.set_timeout_async(worker, delay * 1000)
 
-    def lint(self, view_id, timestamp):
-        """
-        Call back into the main plugin to lint the given view.
+    # We store the last hit_time in any case. This will invalidate the worker
+    # if we have an old one running. The worker needs to debounce and schedule
+    # a new task with a fixed delay accordingly.
+    last_seen[vid] = hit_time
+    return hit_time
 
-        timestamp is used to determine if the view has been modified
-        since the lint was requested.
 
-        """
-        self.callback(view_id, timestamp)
+MIN_DELAY = 0.1
 
-    def get_delay(self, view):
-        """
-        Return the delay between a lint request and when it will be processed.
 
-        If the lint mode is not background, there is no delay. Otherwise, if
-        a "delay" setting is not available in any of the settings, MIN_DELAY is used.
-        """
-        if persist.settings.get('lint_mode') != 'background':
-            return 0
+def get_delay():
+    """Return the delay between a lint request and when it will be processed.
 
-        delay = persist.settings.get('delay', self.MIN_DELAY)
+    If the lint mode is not background, there is no delay. Otherwise, if
+    a "delay" setting is not available in any of the settings, MIN_DELAY is used.
```

Comment:
Delay is guaranteed to be available because it's in the default settings and those are immutable now. The `MIN_DELAY` can go.